### PR TITLE
Add tests for various instructions

### PR DESCRIPTION
**Pull request by leopoldwe**
Adds tests for ddivu, dmult, dmultu, dsllv, dsra, dsrlv, lui, mult, multu, ori, sra, srav, and xori.

Partially solves #240 and includes a test for PR #481.

### CODE REVIEW

**Comment by lupino3** [comment]
Thanks a lot @leopoldwe, that's a lot of additional test coverage! I will review your PR tomorrow.

**Comment by lupino3** [comment]
Hi @leopoldwe, the build seems to be failing. Could you please have a look into it?

**Comment by lupino3** [comment]
(sorry for the delay, I was busy with the release and the subsequent bugs so I honestly forgot to review your change :( )

File: src/test/java/org/edumips64/EndToEndTests.java [hunk 1]
```diff
@@ -421,6 +421,83 @@ public void testSllv() throws Exception {
...
+  }
+  
+  /* Test for instruction ORI */
+  @Test(timeout=2000)
+  public void testOri() throws Exception {
+    runMipsTest("ori.s");
+  }
+
+  /* Test for instruction LUI */
+  @Test(timeout=2000000000)
```
Comment (lupino3) [inline src/test/java/org/edumips64/EndToEndTests.java:479]: Why is this timeout so high?

Comment (leopoldwe) [reply src/test/java/org/edumips64/EndToEndTests.java]: I was doing debugging, so I increased the timeout. I changed it back to 2000.

**Comment by lupino3** [comment]
@leopoldwe I think you need to rebase this PR on top of the master branch, to pick up the fix for the `LUI` instruction which I merged last week.

**Comment by leopoldwe** [comment]
Done

**Comment by lupino3** [comment]
Thanks @leopoldwe. The tests are still failing: https://github.com/EduMIPS64/edumips64/pull/488/checks?check_run_id=1316675812

Do the tests succeed on your machine?

**Comment by leopoldwe** [comment]
@lupino3 They do. I am running them in Visual Studio Code, if that matters. The test programs themselves run correctly if I run them from the GUI.

**Comment by lupino3** [comment]
Thanks, I'll have a look and figure out why they don't work properly in GitHub Actions.

What operating system and Java JDK are you using locally?

**Comment by leopoldwe** [comment]
macOS 10.12.6
JDK 11.0.8

**Comment by lupino3** [comment]
I just tried running the tests locally, and they fail for me, similarly to CI. They fail both by running `./gradlew jar` from the CLI and by running them in VS Code
Screenshot 2020-10-28 at 21 27 25

Both some E2E test and some CliRunnerTests fail.

Now, the odd thing is that CliRunnerTests are failing for `master` as well. Which is odd because they pass on CI, I wonder if that's a problem with Mac OS X...

I'll need to investigate this a bit more. Sorry for the issues!

Can you let me know what is the result of running `./gradlew test` from the CLI?

**Comment by lupino3** [comment]
I found two issues with CLI Runner tests, #499 and #500. This is why CLI Runner tests were failing on my machine.

The issues are independent from the failures I see with tests in this PR. As I mitigate those 2 issues, tests from `master` succeed but the ones from this PR fail.

I'll dig a bit deeper into this.

**Comment by lupino3** [comment]
At least on my local machine, if I comment out `testLui` the test suite works well. If I uncomment this test, many other tests end up failing (but interestingly enough not `testLui`, apparently).

Note that I am doing my test from a 2015 MacBook Air, which means it's more likely that some slower tests may exceed the timeout. So my hunch right now is that 1) `testLui` might time out on slow machines. 2) the test fixture may not deal well with test timeouts.

**Comment by leopoldwe** [comment]
> Configure project :
Running locally

> Task :test

org.edumips64.EndToEndTests > testOri FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testSra FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testNegativeAddress FAILED
    java.lang.AssertionError at ExpectException.java:34

org.edumips64.EndToEndTests > testReadWriteFile FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testCvtdw FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testDdivu FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testDmult FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testDsllv FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testDsrlv FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testMultu FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testBgez FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testMult FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testSrav FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testLdLargeImmediate FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testTerminationInID FAILED
    org.junit.runners.model.TestTimedOutException at EndToEndTests.java:188

org.edumips64.EndToEndTests > testImmediateOverflow FAILED
    java.lang.AssertionError at ExpectException.java:34

org.edumips64.EndToEndTests > testRAWForRTypeFlowControl FAILED
    org.junit.runners.model.TestTimedOutException at EndToEndTests.java:188

org.edumips64.EndToEndTests > testIssue304UI FAILED
    org.junit.runners.model.TestTimedOutException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testHailStone FAILED
    org.junit.runners.model.TestTimedOutException at EndToEndTests.java:187

org.edumips64.EndToEndTests > storeAfterLoad FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testDmultu FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testMemory FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testPrintf FAILED
    org.edumips64.core.MemoryElementNotFoundException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testStrcmp FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.EndToEndTests > testStrlen FAILED
    org.edumips64.core.is.BreakException at EndToEndTests.java:187

org.edumips64.CliRunnerTest > can_run_a_file FAILED
    java.lang.AssertionError at CliRunnerTest.java:35

238 tests completed, 26 failed

> Task :test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///Users/Leopold/Documents/GitHub-Projects/edumips64/build/reports/tests/test/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 59s
5 actionable tasks: 2 executed, 3 up-to-date

**Comment by lupino3** [comment]
I don't think it's the timeout, because even if I set a very large timeout it still fails.

**Comment by lupino3** [comment]
I think this might be a deeper bug in `LUI` or somewhere else. `testLui` passes, but it seems to create problems for other tests.

Which is pretty puzzling to me, since my understanding is that we are re-creating the CPU and all other simulator classes.

Maybe there's a bug affecting a static field somewhere.

**Comment by lupino3** [comment]
Yes. This can be reproduced by doing the following:

1. open `bgez.s`
2. run it. See it succeed (i.e., not running into the `BREAK` instruction)
3. open `lui.s`
4. run it. See it succeed.
5. open `bgez.s` again
6. run it. It now fails!

**Comment by lupino3** [comment]
Yeah, found the bug. LUI modifies a static field of `ALU_IType` and breaks the simulator forever once it's executed for the first time 💀 : https://github.com/EduMIPS64/edumips64/blob/master/src/main/java/org/edumips64/core/is/LUI.java#L45

This is a *huge* bug.

Thanks @leopoldwe for exposing it with your unit tests!

I'll fix it and ask you to please rebase your tests on top of it once it lands on `master`.

**Comment by lupino3** [comment]
Hi @leopoldwe, I merged #502, which should fix the LUI bug.

Could you please rebase your PR on top of `master`? Now the tests should pass.

Thanks a lot for your contribution! Your help was invaluable to identify the broken instruction!

**Comment by leopoldwe** [comment]
Done.

**Review by lupino3 — APPROVED** [review]
Looks great, thanks for contributing all these unit tests!!

**Pull request merged**